// bresenham-zingl/dist/index.mjs
function T(f, a, i, t, M) {
  const h = Math.abs(i - f), o = f < i ? 1 : -1, e = -Math.abs(t - a), l = a < t ? 1 : -1;
  let r = h + e, n;
  for (; ; ) {
    if (M(f, a), n = 2 * r, n >= e) {
      if (f === i) break;
      r += e, f += o;
    }
    if (n <= h) {
      if (a === t) break;
      r += h, a += l;
    }
  }
}
function U(f, a, i, t, M) {
  const h = f < i ? 1 : -1, o = a < t ? 1 : -1;
  let e, l = Math.abs(i - f), r = Math.abs(t - a), n = l * l + r * r, c = n === 0 ? 1 : 1 / Math.sqrt(n);
  for (l *= c, r *= c, n = l - r; ; ) {
    if (M(f, a, Math.abs(n - (l + r)) >> 16), c = n, e = f, 2 * c >= -l) {
      if (f === i) break;
      c + r < 16711680 && M(f, a + o, c + r >> 16), n -= r, f += h;
    }
    if (2 * c <= r) {
      if (a === t) break;
      l - c < 16711680 && M(e + h, a, l - c >> 16), n += l, a += o;
    }
  }
}
function A(f, a, i, t, M, h) {
  let o = Math.abs(i - f), e = f < i ? 1 : -1, l = Math.abs(t - a), r = a < t ? 1 : -1, n = o - l, c, v, u, b = o + l == 0 ? 1 : Math.sqrt(o * o + l * l);
  for (M = (M + 1) / 2; ; ) {
    if (h(
      f,
      a,
      Math.max(0, 255 * (Math.abs(n - o + l) / b - M + 1))
    ), c = n, v = f, 2 * c >= -o) {
      for (c += l, u = a; c < b * M && (t != u || o > l); c += o)
        h(
          f,
          u += r,
          Math.max(0, 255 * (Math.abs(c) / b - M + 1))
        );
      if (f == i) break;
      c = n, n -= l, f += e;
    }
    if (2 * c <= l) {
      for (c = o - c; c < b * M && (i != v || o < l); c += l)
        h(
          v += e,
          a,
          Math.max(0, 255 * (Math.abs(c) / b - M + 1))
        );
      if (a == t) break;
      n += o, a += r;
    }
  }
}
function J(f, a = "assert error") {
  if (!f) throw new Error(a);
}
function w(f, a, i, t, M, h, o, e) {
  var l = f - 2 * i + M, r = a - 2 * t + h, n = f - i, c = a - t, v, u, b;
  J(o >= 0, "width is negative"), n * (M - i) > 0 && (c * (h - t) > 0 && Math.abs(n * r) > Math.abs(c * l) && (f = M, M = n + i, a = h, h = c + t), f == M || o == 1 ? u = (f - i) / l : (b = Math.sqrt(4 * o * o * (f - i) * (M - i) + (M - f) * (M - f)), i < f && (b = -b), u = (2 * o * (f - i) - f + M + b) / (2 * (1 - o) * (M - f))), b = 1 / (2 * u * (1 - u) * (o - 1) + 1), n = (u * u * (f - 2 * o * i + M) + 2 * u * (o * i - f) + f) * b, c = (u * u * (a - 2 * o * t + h) + 2 * u * (o * t - a) + a) * b, v = u * (o - 1) + 1, v *= v * b, o = ((1 - u) * (o - 1) + 1) * Math.sqrt(b), l = Math.floor(n + 0.5), r = Math.floor(c + 0.5), c = (n - f) * (t - a) / (i - f) + a, I(
    f,
    a,
    l,
    Math.floor(c + 0.5),
    l,
    r,
    v,
    e
  ), c = (n - M) * (t - h) / (i - M) + h, t = Math.floor(c + 0.5), f = i = l, a = r), (a - t) * (h - t) > 0 && (a == h || o == 1 ? u = (a - t) / (a - 2 * t + h) : (b = Math.sqrt(4 * o * o * (a - t) * (h - t) + (h - a) * (h - a)), t < a && (b = -b), u = (2 * o * (a - t) - a + h + b) / (2 * (1 - o) * (h - a))), b = 1 / (2 * u * (1 - u) * (o - 1) + 1), n = (u * u * (f - 2 * o * i + M) + 2 * u * (o * i - f) + f) * b, c = (u * u * (a - 2 * o * t + h) + 2 * u * (o * t - a) + a) * b, v = u * (o - 1) + 1, v *= v * b, o = ((1 - u) * (o - 1) + 1) * Math.sqrt(b), l = Math.floor(n + 0.5), r = Math.floor(c + 0.5), n = (i - f) * (c - a) / (t - a) + f, I(
    f,
    a,
    Math.floor(n + 0.5),
    r,
    l,
    r,
    v,
    e
  ), n = (i - M) * (c - h) / (t - h) + M, i = Math.floor(n + 0.5), f = l, a = t = r), I(
    f,
    a,
    i,
    t,
    M,
    h,
    o * o,
    e
  );
}
function I(f, a, i, t, M, h, o, e) {
  var l = M - i, r = h - t, n = f - M, c = a - h, v = f - i, u = a - t, b = v * r + u * l, s = v * r - u * l, d;
  if (J(v * l <= 0 && u * r <= 0, "sign of gradient must not change"), s != 0 && o > 0) {
    if (l * l + r * r > v * v + u * u && (M = f, f -= n, h = a, a -= c, s = -s), v = 2 * (4 * o * l * v + n * n), u = 2 * (4 * o * r * u + c * c), l = f < M ? 1 : -1, r = a < h ? 1 : -1, b = -2 * l * r * (2 * o * b + n * c), s * l * r < 0 && (v = -v, u = -u, b = -b, s = -s), n = 4 * o * (i - f) * r * s + v / 2 + b, c = 4 * o * (a - t) * l * s + u / 2 + b, o < 0.5 && (c > b || n < b)) {
      s = (o + 1) / 2, o = Math.sqrt(o), b = 1 / (o + 1), l = Math.floor(
        (f + 2 * o * i + M) * b / 2 + 0.5
      ), r = Math.floor((a + 2 * o * t + h) * b / 2 + 0.5), n = Math.floor((o * i + f) * b + 0.5), c = Math.floor((t * o + a) * b + 0.5), I(
        f,
        a,
        n,
        c,
        l,
        r,
        s,
        e
      ), n = Math.floor((o * i + M) * b + 0.5), c = Math.floor((t * o + h) * b + 0.5), I(l, r, n, c, M, h, s, e);
      return;
    }
    d = n + c - b;
    do {
      if (e(f, a), f == M && a == h) return;
      i = 2 * d > c ? 1 : 0, t = 2 * (d + u) < -c ? 1 : 0, (2 * d < n || t) && (a += r, c += b, d += n += v), (2 * d > n || i) && (f += l, n += b, d += c += u);
    } while (c <= b && n >= b);
  }
  T(f, a, M, h, e);
}
function X(f, a, i, t, M, h, o, e) {
  var l = M - i, r = h - t, n = f - M, c = a - h, v = f - i, u = a - t, b = v * r + u * l, s = v * r - u * l, d, z, N;
  if (J(
    v * l <= 0 && u * r <= 0
  ), s != 0 && o > 0) {
    if (l * l + r * r > v * v + u * u && (M = f, f -= n, h = a, a -= c, s = -s), v = 2 * (4 * o * l * v + n * n), u = 2 * (4 * o * r * u + c * c), l = f < M ? 1 : -1, r = a < h ? 1 : -1, b = -2 * l * r * (2 * o * b + n * c), s * l * r < 0 && (v = -v, u = -u, s = -s, b = -b), n = 4 * o * (i - f) * r * s + v / 2 + b, c = 4 * o * (a - t) * l * s + u / 2 + b, o < 0.5 && c > n)
      return s = (o + 1) / 2, o = Math.sqrt(o), b = 1 / (o + 1), l = Math.floor(
        (f + 2 * o * i + M) * b / 2 + 0.5
      ), r = Math.floor((a + 2 * o * t + h) * b / 2 + 0.5), n = Math.floor((o * i + f) * b + 0.5), c = Math.floor((t * o + a) * b + 0.5), X(
        f,
        a,
        n,
        c,
        l,
        r,
        s,
        e
      ), n = Math.floor((o * i + M) * b + 0.5), c = Math.floor((t * o + h) * b + 0.5), X(
        l,
        r,
        n,
        c,
        M,
        h,
        s,
        e
      );
    d = n + c - b;
    do {
      if (s = Math.min(n - b, b - c), z = Math.max(n - b, b - c), z += 2 * z * s * s / (4 * z * z + s * s), i = 255 * Math.abs(d - n - c + b) / z, i < 256 && e(f, a, i), N = 2 * d + c < 0) {
        if (a == h) return;
        n - d < z && e(f + l, a, 255 * Math.abs(n - d) / z);
      }
      if (2 * d + n > 0) {
        if (f == M) return;
        d - c < z && e(f, a + r, 255 * Math.abs(d - c) / z), f += l, n += b, d += c += u;
      }
      N && (a += r, c += b, d += n += v);
    } while (c < n);
  }
  U(f, a, M, h, e);
}
function P(f, a, i, t, M) {
  let h = -i, o = 0, e = t * t, l = h * (2 * e + h) + e;
  do
    M(f - h, a + o), M(f + h, a + o), M(f + h, a - o), M(f - h, a - o), e = 2 * l, e >= (h * 2 + 1) * t * t && (l += (++h * 2 + 1) * t * t), e <= (o * 2 + 1) * i * i && (l += (++o * 2 + 1) * i * i);
  while (h <= 0);
  for (; o++ < t; )
    M(f, a + o), M(f, a - o);
}
function y(f, a, i, t, M, h) {
  let o = i * i, e = t * t;
  const l = Math.sin(M);
  let r = (o - e) * l;
  o = Math.sqrt(o - r * l), e = Math.sqrt(e + r * l), i = o + 0.5, t = e + 0.5, r = r * i * t / (o * e), Z(
    f - i,
    a - t,
    f + i,
    a + t,
    4 * r * Math.cos(M),
    h
  );
}
function Z(f, a, i, t, M, h) {
  let o = i - f, e = t - a, l = o * e;
  if (M === 0) return _(f, a, i, t, h);
  l !== 0 && (l = (l - M) / (l + l)), J(l <= 1 && l >= 0, "limit angle to |zd|<=xd*yd"), o = Math.floor(o * l + 0.5), e = Math.floor(e * l + 0.5), I(f, a + e, f, a, f + o, a, 1 - l, h), I(f, a + e, f, t, i - o, t, l, h), I(i, t - e, i, t, i - o, t, 1 - l, h), I(i, t - e, i, a, f + o, a, l, h);
}
function _(f, a, i, t, M) {
  let h = Math.abs(i - f), o = Math.abs(t - a), e = o & 1, l = 4 * (1 - h) * o * o, r = 4 * (e + 1) * h * h, n = l + r + e * h * h, c;
  f > i && (f = i, i += h), a > t && (a = t), a += (o + 1) / 2, t = a - e, h = 8 * h * h, e = 8 * o * o;
  do
    M(i, a), M(f, a), M(f, t), M(i, t), c = 2 * n, c <= r && (a++, t--, n += r += h), (c >= l || 2 * n > r) && (f++, i--, n += l += e);
  while (f <= i);
  for (; a - t <= o; )
    M(f - 1, a), M(i + 1, a++), M(f - 1, t), M(i + 1, t--);
}
function x(f, a, i, t) {
  var M = -i, h = 0, o = 2 - 2 * i;
  do
    t(f - M, a + h), t(f - h, a - M), t(f + M, a - h), t(f + h, a + M), i = o, i <= h && (o += ++h * 2 + 1), (i > M || o > h) && (o += ++M * 2 + 1);
  while (M < 0);
}
function aa(f, a, i, t) {
  var M = -i, h = 0, o, e, l, r = 2 - 2 * i;
  i = 1 - r;
  do
    o = 255 * Math.abs(r - 2 * (M + h) - 2) / i, t(f - M, a + h, o), t(f - h, a - M, o), t(f + M, a - h, o), t(f + h, a + M, o), l = r, e = M, r + h > 0 && (o = 255 * (r - 2 * M - 1) / i, o < 256 && (t(f - M, a + h + 1, o), t(f - h - 1, a - M, o), t(f + M, a - h - 1, o), t(f + h + 1, a + M, o)), r += ++M * 2 + 1), l + e <= 0 && (o = 255 * (2 * h + 3 - l) / i, o < 256 && (t(f - e - 1, a + h, o), t(f - h, a - e - 1, o), t(f + e + 1, a - h, o), t(f + h, a + e + 1, o)), r += ++h * 2 + 1);
  while (M < 0);
}
function fa(f, a, i, t, M, h, o) {
  var e = f - i, l = a - t, r = f - 2 * i + M, n;
  e * (M - i) > 0 && (l * (h - t) > 0 && Math.abs((a - 2 * t + h) / r * e) > Math.abs(l) && (f = M, M = e + i, a = h, h = l + t), r = (f - i) / r, n = (1 - r) * ((1 - r) * a + 2 * r * t) + r * r * h, r = (f * M - i * i) * r / (f - i), e = Math.floor(r + 0.5), l = Math.floor(n + 0.5), n = (t - a) * (r - f) / (i - f) + a, L(f, a, e, Math.floor(n + 0.5), e, l, o), n = (t - h) * (r - M) / (i - M) + h, f = i = e, a = l, t = Math.floor(n + 0.5)), (a - t) * (h - t) > 0 && (r = a - 2 * t + h, r = (a - t) / r, n = (1 - r) * ((1 - r) * f + 2 * r * i) + r * r * M, r = (a * h - t * t) * r / (a - t), e = Math.floor(n + 0.5), l = Math.floor(r + 0.5), n = (i - f) * (r - a) / (t - a) + f, L(f, a, Math.floor(n + 0.5), l, e, l, o), n = (i - M) * (r - h) / (t - h) + M, f = e, i = Math.floor(n + 0.5), a = t = l), L(f, a, i, t, M, h, o);
}
function L(f, a, i, t, M, h, o) {
  var e = M - i, l = h - t, r = f - i, n = a - t, c, v, u, b, s = r * l - n * e;
  if (J(r * e <= 0 && n * l <= 0, "sign of gradient must not change"), e * e + l * l > r * r + n * n && (M = f, f = e + i, h = a, a = l + t, s = -s), s != 0) {
    r += e, r *= e = f < M ? 1 : -1, n += l, n *= l = a < h ? 1 : -1, c = 2 * r * n, r *= r, n *= n, s * e * l < 0 && (r = -r, n = -n, c = -c, s = -s), v = 4 * l * s * (i - f) + r - c, u = 4 * e * s * (a - t) + n - c, r += r, n += n, b = v + u + c;
    do {
      if (o(f, a), f == M && a == h) return;
      t = +(2 * b < v), 2 * b > u && (f += e, v -= c, b += u += n), t && (a += l, u -= c, b += v += r);
    } while (u < 0 && v > 0);
  }
  T(f, a, M, h, o);
}
function ta(f, a, i, t, M, h, o) {
  var e = f - i, l = a - t, r = f - 2 * i + M, n;
  e * (M - i) > 0 && (l * (h - t) > 0 && Math.abs((a - 2 * t + h) / r * e) > Math.abs(l) && (f = M, M = e + i, a = h, h = l + t), r = (f - i) / r, n = (1 - r) * ((1 - r) * a + 2 * r * t) + r * r * h, r = (f * M - i * i) * r / (f - i), e = Math.floor(r + 0.5), l = Math.floor(n + 0.5), n = (t - a) * (r - f) / (i - f) + a, O(f, a, e, Math.floor(n + 0.5), e, l, o), n = (t - h) * (r - M) / (i - M) + h, f = i = e, a = l, t = Math.floor(n + 0.5)), (a - t) * (h - t) > 0 && (r = a - 2 * t + h, r = (a - t) / r, n = (1 - r) * ((1 - r) * f + 2 * r * i) + r * r * M, r = (a * h - t * t) * r / (a - t), e = Math.floor(n + 0.5), l = Math.floor(r + 0.5), n = (i - f) * (r - a) / (t - a) + f, O(f, a, Math.floor(n + 0.5), l, e, l, o), n = (i - M) * (r - h) / (t - h) + M, f = e, i = Math.floor(n + 0.5), a = t = l), O(f, a, i, t, M, h, o);
}
function O(f, a, i, t, M, h, o) {
  var e = M - i, l = h - t, r = f - i, n = a - t, c, v, u, b, s, d = r * l - n * e;
  if (e * e + l * l > r * r + n * n && (M = f, f = e + i, h = a, a = l + t, d = -d), d != 0) {
    r += e, r *= e = f < M ? 1 : -1, n += l, n *= l = a < h ? 1 : -1, c = 2 * r * n, r *= r, n *= n, d * e * l < 0 && (r = -r, n = -n, c = -c, d = -d), v = 4 * l * (i - f) * d + r - c, u = 4 * e * (a - t) * d + n - c, r += r, n += n, b = v + u + c;
    do {
      if (d = Math.min(v + c, -c - u), s = Math.max(v + c, -c - u), s += 2 * s * d * d / (4 * s * s + d * d), o(
        f,
        a,
        255 * Math.abs(b - v - u - c) / s
      ), f == M || a == h) break;
      i = f, d = v - b, t = +(2 * b + u < 0), 2 * b + v > 0 && (b - u < s && o(f, a + l, 255 * Math.abs(b - u) / s), f += e, v -= c, b += u += n), t && (d < s && o(i + e, a, 255 * Math.abs(d) / s), a += l, u -= c, b += v += r);
    } while (u < v);
  }
  U(f, a, M, h, o);
}
function $(f, a, i, t, M, h, o, e, l) {
  var r, n, c, v = 1;
  let u = f < o ? 1 : -1, b = a < e ? 1 : -1, s = -Math.abs(f + i - M - o), d = s - 4 * u * (i - M), z = u * (f - i - M + o), N = -Math.abs(a + t - h - e), B = N - 4 * b * (t - h), S = b * (a - t - h + e), G, k, W, R, j, E, C, q, g, p, H, F = 0.01;
  if (J(
    (i - f) * (M - o) < F && ((o - f) * (i - M) < F || z * z < d * s + F),
    "slope change"
  ), J(
    (t - a) * (h - e) < F && ((e - a) * (t - h) < F || S * S < B * N + F),
    "slope change"
  ), d == 0 && B == 0)
    return u = Math.floor((3 * i - f + 1) / 2), b = Math.floor((3 * t - a + 1) / 2), L(f, a, u, b, o, e, l);
  i = (i - f) * (i - f) + (t - a) * (t - a) + 1, M = (M - o) * (M - o) + (h - e) * (h - e) + 1;
  do {
    G = d * S - z * B, k = d * N - s * B, W = z * N - s * S, p = G * (G + k - 3 * W) + k * k, r = p > 0 ? 1 : Math.sqrt(1 + 1024 / i), G *= r, k *= r, W *= r, p *= r * r, E = 9 * (G + k + W) / 8, R = 8 * (d - B), q = 27 * (8 * G * (S * S - B * N) + p * (B + 2 * S + N)) / 64 - B * B * (E - B), g = 27 * (8 * G * (z * z - d * s) - p * (d + 2 * z + s)) / 64 - d * d * (E + d), j = 3 * (3 * G * (3 * S * S - B * B - 2 * B * N) - B * (3 * k * (B + S) + B * R)) / 4, C = 3 * (3 * G * (3 * z * z - d * d - 2 * d * s) - d * (3 * k * (d + z) + d * R)) / 4, E = d * B * (6 * G + 6 * k - 3 * W + R), k = B * B, R = d * d, E = 3 * (E + 9 * r * (R * S * N - z * s * k) - 18 * z * S * G) / 8, p < 0 && (q = -q, g = -g, j = -j, C = -C, E = -E, k = -k, R = -R), G = 6 * B * k, k = -6 * d * k, W = 6 * B * R, R = -6 * d * R, q += E, p = q + g, g += E;
    a: for (H = E, n = c = r; f != o && a != e; ) {
      l(f, a);
      do {
        if (q > H || g < H)
          break a;
        t = 2 * p - g, 2 * p >= q && (n--, p += q += j, g += E += k, C += W, j += G), t <= 0 && (c--, p += g += C, q += E += W, j += k, C += R);
      } while (n > 0 && c > 0);
      2 * n <= r && (f += u, n += r), 2 * c <= r && (a += b, c += r), H == E && q < 0 && g > 0 && (H = F);
    }
    j = f, f = o, o = j, u = -u, z = -z, C = a, a = e, e = C, b = -b, S = -S, i = M;
  } while (v--);
  T(
    f,
    a,
    o,
    e,
    l
  );
}
function m(f, a, i, t, M, h, o, e, l) {
  let r, n, c, v = 1, u = f < o ? 1 : -1, b = a < e ? 1 : -1, s = -Math.abs(f + i - M - o), d = s - 4 * u * (i - M), z = u * (f - i - M + o), N = -Math.abs(a + t - h - e), B = N - 4 * b * (t - h), S = b * (a - t - h + e), G, k, W, R, j, E, C, q, g, p, H, F, D, V;
  const K = 0.01;
  if (J(
    (i - f) * (M - o) < K && ((o - f) * (i - M) < K || z * z < d * s + K)
  ), J(
    (t - a) * (h - e) < K && ((e - a) * (t - h) < K || S * S < B * N + K)
  ), d === 0 && B === 0)
    return u = Math.floor((3 * i - f + 1) / 2), b = Math.floor((3 * t - a + 1) / 2), O(f, a, u, b, o, e, l);
  i = (i - f) * (i - f) + (t - a) * (t - a) + 1, M = (M - o) * (M - o) + (h - e) * (h - e) + 1;
  do {
    G = d * S - z * B, k = d * N - s * B, W = z * N - s * S, V = 4 * G * W - k * k, p = G * (G + k - 3 * W) + k * k, r = p > 0 ? 1 : Math.sqrt(1 + 1024 / i), G *= r, k *= r, W *= r, p *= r * r, E = 9 * (G + k + W) / 8, R = 8 * (d - B), q = 27 * (8 * G * (S * S - B * N) + p * (B + 2 * S + N)) / 64 - B * B * (E - B), g = 27 * (8 * G * (z * z - d * s) - p * (d + 2 * z + s)) / 64 - d * d * (E + d), j = 3 * (3 * G * (3 * S * S - B * B - 2 * B * N) - B * (3 * k * (B + S) + B * R)) / 4, C = 3 * (3 * G * (3 * z * z - d * d - 2 * d * s) - d * (3 * k * (d + z) + d * R)) / 4, E = d * B * (6 * G + 6 * k - 3 * W + R), k = B * B, R = d * d, E = 3 * (E + 9 * r * (R * S * N - z * s * k) - 18 * z * S * G) / 8, p < 0 && (q = -q, g = -g, j = -j, C = -C, E = -E, k = -k, R = -R), G = 6 * B * k, k = -6 * d * k, W = 6 * B * R, R = -6 * d * R, q += E, p = q + g, g += E;
    let Q = !1;
    a: for (n = c = r; f !== o && a !== e; ) {
      t = Math.min(Math.abs(E - q), Math.abs(g - E)), D = Math.max(Math.abs(E - q), Math.abs(g - E)), D = r * (D + 2 * D * t * t / (4 * D * D + t * t)), t = 255 * Math.abs(p - (r - n + 1) * q - (r - c + 1) * g + r * E) / D, t < 256 && l(f, a, t), H = Math.abs(p - (r - n + 1) * q + (c - 1) * g), F = Math.abs(p + (n - 1) * q - (r - c + 1) * g), h = a;
      do {
        if (V >= -K && (q + j > E || g + C < E)) {
          Q = !0;
          break a;
        }
        if (t = 2 * p + q, 2 * p + g > 0)
          n--, p += q += j, g += E += k, C += W, j += G;
        else if (t > 0) {
          Q = !0;
          break a;
        }
        t <= 0 && (c--, p += g += C, q += E += W, j += k, C += R);
      } while (n > 0 && c > 0);
      2 * c <= r && (F < D && l(f + u, a, 255 * F / D), a += b, c += r), 2 * n <= r && (H < D && l(f, h + b, 255 * H / D), f += u, n += r);
    }
    Q && (2 * p < g && 2 * c <= r + 2 && (F < D && l(f + u, a, 255 * F / D), a += b), 2 * p > q && 2 * n <= r + 2 && (H < D && l(f, h + b, 255 * H / D), f += u), j = f, f = o, o = j, u = -u, z = -z, C = a, a = e, e = C, b = -b, S = -S, i = M);
    break;
  } while (v--);
  U(
    Math.floor(f),
    Math.floor(a),
    Math.floor(o),
    Math.floor(e),
    l
  );
}
function Y(f, a, i, t, M, h, o, e, l, r) {
  let n = 0, c = 0, v = f + i - M - o, u = v - 4 * (i - M), b = f - i - M + o, s = b + 4 * (i + M), d = a + t - h - e, z = d - 4 * (t - h), N = a - t - h + e, B = N + 4 * (t + h);
  var S = f, G, k, W, R = a, j, E, C;
  let q = b * b - u * v, g;
  const p = [0, 0, 0, 0, 0];
  for (u == 0 ? Math.abs(v) < 2 * Math.abs(b) && (p[n++] = v / (2 * b)) : q > 0 && (g = Math.sqrt(q), q = (b - g) / u, Math.abs(q) < 1 && (p[n++] = q), q = (b + g) / u, Math.abs(q) < 1 && (p[n++] = q)), q = N * N - z * d, z == 0 ? Math.abs(d) < 2 * Math.abs(N) && (p[n++] = d / (2 * N)) : q > 0 && (g = Math.sqrt(q), q = (N - g) / z, Math.abs(q) < 1 && (p[n++] = q), q = (N + g) / z, Math.abs(q) < 1 && (p[n++] = q)), c = 1; c < n; c++)
    (q = p[c - 1]) > p[c] && (p[c - 1] = p[c], p[c] = q, c = 0);
  for (q = -1, p[n] = 1, c = 0; c <= n; c++)
    g = p[c], G = (q * (q * b - 2 * v) - g * (q * (q * u - 2 * b) + v) + s) / 8 - S, j = (q * (q * N - 2 * d) - g * (q * (q * z - 2 * N) + d) + B) / 8 - R, k = (g * (g * b - 2 * v) - q * (g * (g * u - 2 * b) + v) + s) / 8 - S, E = (g * (g * N - 2 * d) - q * (g * (g * z - 2 * N) + d) + B) / 8 - R, S -= W = (g * (g * (3 * b - g * u) - 3 * v) + s) / 8, R -= C = (g * (g * (3 * N - g * z) - 3 * d) + B) / 8, o = Math.floor(W + 0.5), e = Math.floor(C + 0.5), S != 0 && (G *= S = (f - o) / S, k *= S), R != 0 && (j *= R = (a - e) / R, E *= R), (f != o || a != e) && l(f, a, f + G, a + j, f + k, a + E, o, e, r), f = o, a = e, S = W, R = C, q = g;
}
function ra(f, a, i, t, M, h, o, e, l) {
  Y(
    f,
    a,
    i,
    t,
    M,
    h,
    o,
    e,
    $,
    l
  );
}
function oa(f, a, i, t, M, h, o, e, l) {
  Y(
    f,
    a,
    i,
    t,
    M,
    h,
    o,
    e,
    m,
    l
  );
}
export {
  x as circle,
  aa as circleAA,
  ra as cubicBezier,
  oa as cubicBezierAA,
  $ as cubicBezierSegment,
  m as cubicBezierSegmentAA,
  P as ellipse,
  _ as ellipseRect,
  T as line,
  U as lineAA,
  A as lineWidth,
  fa as quadBezier,
  ta as quadBezierAA,
  L as quadBezierSegment,
  O as quadBezierSegmentAA,
  w as quadRationalBezier,
  I as quadRationalBezierSegment,
  X as quadRationalBezierSegmentAA,
  y as rotatedEllipse,
  Z as rotatedEllipseRect
};
